Improve verbose and debug logging level messaging in web cmdlets

JustinGrote opened the pull request:
# PR Summary
- Moves debug message processing to separate methods to reduce `GetResponse` cognitive load
- Verbose messages redesigned to a one-line summary suited for watching the HTTP transactional flow of a script/module/etc.
- Debug messages redesigned to show query, header and body content, and intelligently hide binary content to prevent trashing the console/logs
- Messages are now conditionally generated based on Preference setting for performance.
- Humanize the content payload byte length (thanks @kilasuit)
- For Invoke-Restmethod, show the response encoding detail as a debug message

## Example
![image](https://github.com/user-attachments/assets/d0101095-1231-46cc-a0e9-0a8e898ef329)

## WG Review Response
> The Cmdlet WG discussed this. Since the introduction of Windows PowerShell, binary cmdlets emit both the Verbose and the Debug stream when -Debug is used. This introduces some natural redundancy when i.e. the verbose stream summarizes general operations whereas the debug stream provides more detailed info. Since both streams serve different purposes and audiences, neither one should be omitted just because of potential redundancy. As a workaround, either stream can be redirected to $null if not wanted. We continue to agree that the Debugstream is appropriate for technical details and sensitive information, whereas the Verbose stream is best for summary information.

Agreed. I have defined the intent of the messages to be as such:
**Verbose:** One-Liner showing the basics of what is happening, very useful for concisely tracing program flow. Sensitive information has been redacted at this level in general per [Security WG comments](https://github.com/PowerShell/PowerShell/issues/23879#issuecomment-2243692063) but URL is preserved for troubleshooting purposes and basic logging flow if necessary.

**Debug:** Includes all sensitive information including sensitive headers which may include authentication tokens, etc. as the purpose is for debugging, not logging, and specifying DebugPreference or Debug is considered an "opt-in" for this information.

## PR Context
Closes #25492

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
  - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge. If this PR is a work in progress, please open this as a [Draft Pull Request and mark it as Ready to Review when it is ready to merge](https://docs.github.com/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests#draft-pull-requests).
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
  - [x] None
  - **OR**
  - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.5/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
    - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
  - [ ] Not Applicable
  - **OR**
  - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
  - [ ] N/A or can only be tested interactively
  - **OR**
  - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)

JustinGrote committed:
WebRequest Cmdlets: Improve Verbose and Debug Log

- Adds a clean single-line format for Verbose and a more detailed output for Debug
- Messages are optimized to only be generated if Verbose or Debug have been specified
- Byte lengths are humanized for better readability
- Intelligently decodes only non-binary request/response bodies
Closes #25492

JustinGrote commented:
@iSazonov ready to review!

iSazonov commented:
I wonder how many code you update. 😄 I expect you add only content output.

I like you create helper methods. But now output is too large and we must generate the output only it is requested.
I don't find method(s) to check the condition and we could create it here.
For start see how  `internal void WriteDebug(DebugRecord record, bool overrideInquire = false)` does the check.
We can get ((MshCommandRuntime)this.CommandRuntime).DebugPreference and reuse code from `internal bool WriteHelper_ShouldWrite`
(All in src\System.Management.Automation\engine\MshCommandRuntime.cs)

The same for verbose. But I wouldn't waste time on verbose right now at all.

Also I'd prefer to use standard TosString() for request and response. Why do we create custom implementation?

JustinGrote commented:
@iSazonov I had this discussion with @SeeminglyScience on discord and it is noted there isn't an ideal reliable way to check preference (if you check the first commit I was testing for it prior) and he recommended not establishing it as a pattern but rather this should be something the runtime provides as part of the PSCmdlet abstract class. 

In testing the code has a non-visible impact even on large payloads if Verbose/Debug aren't specified.

I moved away from the Request/Response message `toString` in order to match the format as noted, and reduce redundant information between Verbose and Debug. I put the user experience in favor of the direct implementation. We can revert that if required, but I would lobby for this, especially since I try/catch potential edge cases (haven't hit any so far in testing)

iSazonov commented:
@JustinGrote Yes, we haven't std methods for the checks. I think it's important to have them because later we can add even more output (for example certificates)
One way is to create them. So I open #25514. If you could ask him in Discord to review it, that would be great.
Another way is to use simple workaround only for interactive scenario. We could check switches by more reliable way than you did - `Debug.IsPresent`/`Verbose.IsPresent`. We could use this if my PR will be blocked/rejected.

JustinGrote commented:
Sounds good. Do we want to merge this PR as-is for the next preview and add the conditional logic later, or wait for #25514?

iSazonov commented:
Let's wait for what Patrick says. If that PR is not approved, then let's look at switch checks. What I definitely would not like is to create these strings without a condition. 
We can discuss the details for now.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -1433,6 +1413,183 @@ internal virtual HttpResponseMessage GetResponse(HttpClient client, HttpRequestM
+
+                if (request.Content is not null)
+                {
+                    debugBuilder
+                    .AppendLine(DebugHeaderPrefix + "BODY")
+                    .AppendLine(request.Content switch
+                    {
+                        StringContent stringContent => stringContent
+                        .ReadAsStringAsync(_cancelToken.Token)
+                        .GetAwaiter().GetResult(),
+                        MultipartFormDataContent multipartContent => "=> Multipart Form Content"
+                        + Environment.NewLine
+                        + multipartContent.ReadAsStringAsync(_cancelToken.Token)
+                        .GetAwaiter().GetResult(),
+                        ByteArrayContent byteContent => InFile is not null
+                        ? "[Binary content: "
+                        + GetFriendlyContentLength(byteContent.Headers.ContentLength)
+                        + "]"
+                    : byteContent.ReadAsStringAsync(_cancelToken.Token).GetAwaiter().GetResult(),
+                    StreamContent streamContent => "[Stream content: " + GetFriendlyContentLength(streamContent.Headers.ContentLength) + "]",
+                    _ => "[Unknown content type]",
+                });
```
It looks terrible. Could you have done a better code formatting?

JustinGrote replied:
OK I don't know what happened here, it didn't look terrible before, the indentations were previously present, will fix.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -1433,6 +1413,183 @@ internal virtual HttpResponseMessage GetResponse(HttpClient client, HttpRequestM
+        {
+            try
+            {
+                StringBuilder verboseBuilder = new();
```
The same.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -1433,6 +1413,183 @@ internal virtual HttpResponseMessage GetResponse(HttpClient client, HttpRequestM
+            }
+        }
+
+        private void WriteWebResponseDebugInfo(HttpResponseMessage response)
```
The same comments for the method.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -1433,6 +1413,183 @@ internal virtual HttpResponseMessage GetResponse(HttpClient client, HttpRequestM
+
+                    if (IsTextBasedContentType(GetContentType(response)))
+                    {
+                        debugBuilder.AppendLine(
+                        response.Content.ReadAsStringAsync(_cancelToken.Token)
+                        .GetAwaiter().GetResult());
```
```suggestion
                        debugBuilder.AppendLine(
                            response.Content.ReadAsStringAsync(_cancelToken.Token).GetAwaiter().GetResult());
```

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -1822,7 +1979,7 @@ private static string FormatErrorMessage(string error, string contentType)
             try
             {
-                if (ContentHelper.IsXml(contentType))
+                if (IsXml(contentType))
```
I'd prefer old style.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -300,6 +300,7 @@ function ExecuteRestMethod {
     $verbosePreferenceSave = $VerbosePreference
     $VerbosePreference = 'Continue'
     try {
+        $ENV:__PWSH_TESTOUTPUT = 'true'
```
Why do we need this?

JustinGrote replied:
Noted above. If it really bothers you what I can do is include the encoding information into the debug output and adjust this test to look for it that way.

iSazonov reviewed (commented):
Mainly style comments.

JustinGrote commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -7,6 +7,7 @@
 using System.Collections.ObjectModel;
 using System.Globalization;
 using System.IO;
+using System.Linq;
```
Can you say where that guideline is? [Linq is used all over the codebase...](https://github.com/search?q=repo%3APowerShell%2FPowerShell%20using%20system.linq&type=code)

That said, yes, I'm just using it to sort the imports.

iSazonov replied:
https://github.com/PowerShell/PowerShell/blob/master/docs/dev-process/coding-guidelines.md#performance-considerations

JustinGrote replied:
I will argue that logic seems held over from .NET framework days, modern Linq is very fast and doing a foreach iteration with ordering is probably *less* performant than doing an OrderBy as I have to allocate it to a separate list to get what I want. That said I accept that these conditions are here and will remove the sorting altogether to simplify that fairly ugly part of the code, we can always sort it later if there is greater demand for it.

iSazonov replied:
.Net team optimized Linq for popular scenarios (but not all) so the recommendation is not critical today. And there are examples of using Linq in PowerShell. It is used when it is justified.

JustinGrote commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -23,6 +24,7 @@
 using System.Threading;
 using System.Threading.Tasks;
 using System.Xml;
+using static Microsoft.PowerShell.Commands.ContentHelper;
```
Can do, I did this because the prefix made some parts of the code overly convoluted in my opinion but will do that.

JustinGrote commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -1433,6 +1413,183 @@ internal virtual HttpResponseMessage GetResponse(HttpClient client, HttpRequestM
+            {
+                StringBuilder verboseBuilder = new();
+
+                // "Redact" the query string from verbose output, the details will be visible in Debug output
```
The URL is displayed as part of the Verbose message, the only part redacted is the query string, so that verbose output doesn't show anything sensitive e.g. https://myazurefunction?code=ASECRETCODETOACCESS

So -Verbose is still "safe" as it is used frequently, and you have to use -Debug to "opt in" to seeing potentially sensitive stuff is the thinking.

JustinGrote commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -1433,6 +1413,183 @@ internal virtual HttpResponseMessage GetResponse(HttpClient client, HttpRequestM
+        {
+            try
+            {
+                StringBuilder verboseBuilder = new();
```
Sure, that's reasonable, even if it doesn't get to that size, the hit is negligible.

JustinGrote commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -1433,6 +1413,183 @@ internal virtual HttpResponseMessage GetResponse(HttpClient client, HttpRequestM
+                }
+                foreach (var header in allHeaders.OrderBy(h => h.Key))
+                {
+                    debugBuilder.AppendLine($"{header.Key}: {string.Join(", ", header.Value)}");
```
TIL! Will fix.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -1433,6 +1413,183 @@ internal virtual HttpResponseMessage GetResponse(HttpClient client, HttpRequestM
+                        allHeaders.AddRange(headerSource);
+                    }
+                }
+                foreach (var header in allHeaders.OrderBy(h => h.Key))
```
> doesn't matter per RFC spec

Formally yes. In practice, someone can write simple code for their needs (it can be a PowerShell script) that depends on their order.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -1280,40 +1287,13 @@ internal virtual HttpResponseMessage GetResponse(HttpClient client, HttpRequestM
-
-                    WriteDebug(reqDebugMsg);
+                    WriteWebRequestVerboseInfo(request);
+                    WriteWebRequestDebugInfo(request);
```
I meant that, for example, some streams need to be rewound after reading (otherwise we will get an empty one) and for some it is impossible. In other words, the debugging code can disrupt normal operation.

By the way, the new internal methods are already merged, and you can use them to write debug under the condition.

JustinGrote replied:
@iSazonov understood, but in this case we get very specific Content streams that are resumable and rewindable, so unless something radical happens in the API surface where invoke-restmethod lets you supply *any* stream, it should be safe. I will still gate it behind the conditionals to be safe.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/InvokeRestMethodCommand.Common.cs:
```diff
@@ -110,7 +110,10 @@ internal override void ProcessResponse(HttpResponseMessage response)
 
                     // NOTE: Tests use this verbose output to verify the encoding.
-                    WriteVerbose($"Content encoding: {encodingVerboseName}");
+                    if (Environment.GetEnvironmentVariable("__PWSH_TESTOUTPUT") == "true")
```
It looks like useful information for verbose or debug output. We can make it mandatory. So we can delete this env variable and update the test accordingly.

JustinGrote replied:
Will do

JustinGrote commented:
@iSazonov thank you for the review, I will incorporate the recently merged PR and your feedback later

JustinGrote commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/ContentHelper.Common.cs:
```diff
@@ -133,10 +137,15 @@ internal static bool IsXml([NotNullWhen(true)] string? contentType)
             return isXml;
         }
 
+        internal static bool IsTextBasedContentType([NotNullWhen(true)] string? contentType) =>
+            IsText(contentType)
+            || IsJson(contentType)
+            || IsXml(contentType);
```
I prefer the verticality so that it is more extensible but I see your argument as well and will adjust.

JustinGrote commented:
@iSazonov feedback incorporated and it is now conditional using the new internal `IsEnabled` checks

JustinGrote commented:
@iSazonov changes requested implemented (I went with a nested foreach for the headers to make that easier to read), and made CodeFactor happy. I also squash rebased to a single commit so this should be mergeable now. Let me know any further issues.

JustinGrote commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/InvokeRestMethodCommand.Common.cs:
```diff
@@ -99,18 +99,20 @@ internal override void ProcessResponse(HttpResponseMessage response)
-                        encodingVerboseName = encoding.HeaderName;
+                        // NOTE: This is a getter method that may possibly throw a NotSupportedException exception,
+                        // hence the try/catch
+                        encodingName = encoding.HeaderName;
```
Only because HeaderName is what was there before. Let me investigate if it's viable.

JustinGrote replied:
EncodingName is the "friendly" name of the encoding, HeaderName is the one that you would need to know if you need to specify a specific mime type. I feel this is still the correct one to return as it requires no "translation" by the user if they are trying to submit a specific type of encoding to an API.

Examples:
https://learn.microsoft.com/en-us/dotnet/api/system.text.encoding.encodingname?view=net-9.0#examples

JustinGrote commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -1437,13 +1430,181 @@ internal virtual void UpdateSession(HttpResponseMessage response)
 
         internal static TimeSpan ConvertTimeoutSecondsToTimeSpan(int timeout) => timeout > 0 ? TimeSpan.FromSeconds(timeout) : Timeout.InfiniteTimeSpan;
 
+        private void WriteWebRequestVerboseInfo(HttpRequestMessage request)
```
The `WriteVerbose` PSCmdlet references prevent this, since they are instance local. I could add the PSCmdlet as a parameter, but I don't really see the value in that, and it could be dangerous across runspaces if it were ever used improperly.

JustinGrote commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -1437,13 +1430,181 @@ internal virtual void UpdateSession(HttpResponseMessage response)
+            }
+        }
+
+        private void WriteWebRequestDebugInfo(HttpRequestMessage request)
```
See above, `WriteDebug` reference prevents this without adding PSCmdlet as a parameter.

JustinGrote replied:
See other comment, I don't think this is worth it.

JustinGrote commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -1437,13 +1430,181 @@ internal virtual void UpdateSession(HttpResponseMessage response)
+                            .GetAwaiter().GetResult(),
+                        ByteArrayContent byteContent => InFile is not null
+                            ? "[Binary content: "
+                                + ContentHelper.GetFriendlyContentLength(byteContent.Headers.ContentLength)
```
I'll go the combination route.

JustinGrote commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -1437,13 +1430,181 @@ internal virtual void UpdateSession(HttpResponseMessage response)
+            }
+        }
+
+        private void WriteWebResponseVerboseInfo(HttpResponseMessage response)
```
To make it static would require a separate helper method for the try/catch, and that helper method would need to be supplied the PSCmdlet. I think that obscures things more than it helps, as the static method isn't intended to be used outside the webcmdlet classes.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -1478,6 +1668,7 @@ private static Uri CheckProtocol(Uri uri)
 
             return uri.IsAbsoluteUri ? uri : new Uri("http://" + uri.OriginalString);
         }
+#nullable disable
```
```suggestion
#nullable restore
```

JustinGrote replied:
Oops good catch!

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -1280,40 +1285,13 @@ internal virtual HttpResponseMessage GetResponse(HttpClient client, HttpRequestM
-                        response.ToString());
-
-                    WriteDebug(resDebugMsg);
+                    WriteMessageIfPreferenceEnabled(() => GetWebResponseVerboseInfoString(response), isDebug: false);
+                    WriteMessageIfPreferenceEnabled(() => GetWebResponseDebugInfoString(response, _cancelToken.Token), isDebug: true);
```
Oh, I don't want us to create these delegates on the fly. Earlier, you convinced me that static is not worth it.

JustinGrote replied:
Wasn't sure if I did. I'll revert it.

JustinGrote replied:
Harder than I thought to deduplicate that and not have it be a mess so will full revert.

JustinGrote commented:
@iSazonov static refactor reverted. CodeFactor however is now complaining about complex methods I didn't touch. I tried a rebase/reset and confirmed I didn't touch those methods as far as I can tell. Any idea what's going on? Is it the nullable maybe?

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/ContentHelper.Common.cs:
```diff
@@ -21,8 +21,15 @@ internal static class ContentHelper
 
+        internal static string GetFriendlyContentLength(long? length) =>
+            length.HasValue
+            ? $"{length.Value.Bytes().Humanize()} ({length.Value.ToString("#,0").Replace(",", " ")} bytes)"
```
I guess Humanize uses current culture. So ToString should do the same.
```suggestion
            ? $"{length.Value.Bytes().Humanize()} ({length.Value.ToString("#,0")} bytes)"
```

iSazonov reviewed (dismissed):
Please update the PR description to follow latest changes.

JustinGrote commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -1437,13 +1430,210 @@ internal virtual void UpdateSession(HttpResponseMessage response)
+
+                    foreach (var header in headerSet)
+                    {
+                        debugBuilder.AppendLine($"{header.Key}: {string.Join(", ", header.Value)}");
```
Good find!

JustinGrote commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/InvokeRestMethodCommand.Common.cs:
```diff
@@ -99,18 +99,20 @@ internal override void ProcessResponse(HttpResponseMessage response)
-                    // NOTE: Tests use this verbose output to verify the encoding.
-                    WriteVerbose($"Content encoding: {encodingVerboseName}");
+                    // NOTE: Tests use this debug output to verify the encoding.
+                    WriteDebug($"WebResponse content encoding: {encodingName}");
```
It's just debug and only shows up for restmethod, so I'll just add friendlyname, encodingname, and encoding page.

JustinGrote replied:
This makes the test more complicated FYI

JustinGrote committed:
Incorporate latest @iSazonov feedback

JustinGrote commented:
@iSazonov incorporated feedback. Codefactor still complaining about complex methods I didn't touch.

iSazonov commented:
> @iSazonov incorporated feedback. Codefactor still complaining about complex methods I didn't touch.

@JustinGrote Please ignore such issues.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/InvokeRestMethodCommand.Common.cs:
```diff
@@ -99,20 +99,25 @@ internal override void ProcessResponse(HttpResponseMessage response)
 
                     // NOTE: Tests use this debug output to verify the encoding.
-                    WriteDebug($"WebResponse content encoding: {encodingName}");
+                    if (encodingHeaderName != string.Empty && encodingName != string.Empty)
+                    {
+                        WriteDebug($"WebResponse content encoding: {encodingHeaderName} ({encodingName}) CodePage: {encodingPage}");
+                    }
```
I think we should write in any case without the condition. And replace `string.Empty` with "unknown" or "-".

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -297,37 +297,37 @@ function ExecuteRestMethod {
-            if ($result.Encoding -eq $null) {
-                throw "Encoding not found in verbose output. Lines: $($result.Verbose.Count) Content:$($result.Verbose)"
+            if ($null -eq $result.Encoding) {
+                throw "Encoding not found in verbose output. Lines: $($result.Debug.Count) Content:$($result.Debug)"
```
Now it is debug output.

JustinGrote committed:
Always emit content encoding message

JustinGrote committed:
Update message for debug

JustinGrote committed:
DebugHeaderPrefix Consistency

iSazonov reviewed (approved):
Code LGTM. Waiting community feedback...

JustinGrote commented:
I'll investigate the test failures.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -1280,40 +1285,27 @@ internal virtual HttpResponseMessage GetResponse(HttpClient client, HttpRequestM
-                        request.ToString());
+                    if (IsWriteVerboseEnabled())
+                    {
+                        WriteWebRequestVerboseInfo(request);
```
I guess we should use `currentRequest` here to fix tests.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -1280,40 +1285,27 @@ internal virtual HttpResponseMessage GetResponse(HttpClient client, HttpRequestM
-                    WriteDebug(reqDebugMsg);
+                    if (IsWriteDebugEnabled())
+                    {
+                        GetWebRequestDebugInfo(request);
```
The same.

iSazonov reviewed (commented):
I see `request` is disposed in failed tests.

JustinGrote committed:
Replace request with currentRequest

JustinGrote committed:
Fix inconsistent method name

JustinGrote commented:
@iSazonov good catch, that was it. Also fixed an inconsistent `Write` method name

JustinGrote commented:
@iSazonov I rebased to latest master and I think that made codefactor mad, you probably need to exclude whatever you did previously for the complex methods.

iSazonov commented:
> I rebased to latest master and I think that made codefactor mad, you probably need to exclude whatever you did previously for the complex methods.

Line 1416 `currentRequest.Dispose();`

iSazonov commented:
@JustinGrote I see some tests still hang ...

JustinGrote commented:
@iSazonov I was having trouble getting my local testing to work specifically for webcmdlets last night but I will try again this evening (US Pacific time)

JustinGrote committed:
Test: Adjust to test for codepage (more deterministic)

JustinGrote commented:
@iSazonov tests fixed, the `-Debug` parameter was causing inquire and since Pester is run interactively, was waiting on the dialog, switched to using debugpreference.

Also adjusted encoding test to match on Codepage since we are providing it anyways, should be a much less brittle test.

iSazonov commented:
@JustinGrote Still hang.  I guess VerbosePreference is inquire.

JustinGrote commented:
Probably related to this, I couldn't get it to reproduce in Linux but let me do it in Windows.
![image](https://github.com/user-attachments/assets/7634ff05-e029-49ca-afe4-d97bd6110b87)

JustinGrote committed:
Remove Unused do/while loop that caused test deadlock.

All retries are done with recursive GetResponse added by #19229.

The do loop was a leftover from the old way of doing things, it never actually gets triggered, and redirects previously had internal GetResponse that bypassed it, but when I unified the logic to unify the verbose messages, it now follows the flow "properly" and got stuck in the do/while instead of return the response, causing it to reuse the request and trigger an exception.

JustinGrote committed:
Add back OutFile to Verbose Output

Fixes test "Invoke-WebRequest -PassThru -OutFile -Verbose File Name reflects the downloaded file name"

JustinGrote committed:
Revert "Remove Unused do/while loop that caused test deadlock."

This reverts commit f4f446fe1bad6adb647279083f98875b665bbbe4.

JustinGrote committed:
Remove misplaced verbosebuilder

JustinGrote commented:
This is still active, there's an issue somewhere in the MaximumRetry code where a request is getting reused, I haven't been able to track it down yet and need to step thru the commits to find where it's happening.

JustinGrote committed:
Revert Accidental totalrequests-- change

JustinGrote commented:
@iSazonov I found the problem. Please approve for tests to run, it should fully pass now and be mergeable.

JustinGrote commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -1280,40 +1285,27 @@ internal virtual HttpResponseMessage GetResponse(HttpClient client, HttpRequestM
+                    // if (IsWriteDebugEnabled())
+                    // {
+                    //     WriteWebRequestDebugInfo(currentRequest);
+                    // }
```
@iSazonov thank you, did it in a rush while finding the issue...

JustinGrote committed:
Revert disabling request verbose/debug info

JustinGrote committed:
Minor formatting and spacing cleanup

JustinGrote committed:
make body size response message consistent with request

iSazonov reviewed (approved):
LGTM with one notice: perhaps it makes sense to remove `body` from debug output because it can be too large..

iSazonov commented:
/azp run PowerShell-CI-linux-packaging,PowerShell-Windows-Packaging-CI

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -1437,13 +1429,206 @@ internal virtual void UpdateSession(HttpResponseMessage response)
+                    else
+                    {
+                        string friendlyContentLength = ContentHelper.GetFriendlyContentLength(
+                            response.Content?.Headers?.ContentLength);
```
Header is not null here.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -1437,13 +1429,206 @@ internal virtual void UpdateSession(HttpResponseMessage response)
+                }
+
+                long? requestContentLength = request.Content?.Headers?.ContentLength;
+                if (requestContentLength is not null)
+                {
+                    verboseBuilder.Append($" with body size {ContentHelper.GetFriendlyContentLength(requestContentLength)}");
+                }
+                if (OutFile is not null)
```
Style: new line is needed after brace.
